src/main/java/org/embulk/input/zendesk/services/ZendeskChatService.java

```diff
@@ -0,0 +1,281 @@
+package org.embulk.input.zendesk.services;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import org.embulk.config.TaskReport;
+import org.embulk.input.zendesk.RecordImporter;
+import org.embulk.input.zendesk.ZendeskInputPlugin;
+import org.embulk.input.zendesk.clients.ZendeskRestClient;
+
+import org.embulk.input.zendesk.utils.ZendeskConstants;
+import org.embulk.input.zendesk.utils.ZendeskDateUtils;
+import org.embulk.input.zendesk.utils.ZendeskUtils;
+import org.embulk.spi.Exec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.IntStream;
+
+public class ZendeskChatService implements ZendeskService
+{
+    private static final Logger logger = LoggerFactory.getLogger(ZendeskChatService.class);
+
+    private ZendeskRestClient zendeskRestClient;
+
+    protected ZendeskInputPlugin.PluginTask task;
+
+    private static final int MAXIMUM_TOTAL_RECORDS = 10000;
+
+    private static final int MAXIMUM_TOTAL_PAGES = 250;
+
+    private static final int MAXIMUM_RECORDS_PER_PAGE = 40;
+
+    public ZendeskChatService(final ZendeskInputPlugin.PluginTask task)
+    {
+        this.task = task;
+    }
+
+    @Override
+    public boolean isSupportIncremental()
+    {
+        return true;
+    }
+
+    @Override
+    public TaskReport addRecordToImporter(final int taskIndex, final RecordImporter recordImporter)
+    {
+        final TaskReport taskReport = Exec.newTaskReport();
+        String startTime = getStartTime();
+        String endTime = getEndTime();
+
+        if (Exec.isPreview()) {
+            fetchData(startTime, endTime, 1, recordImporter);
+            return taskReport;
+        }
+
+        while (true) {
+            String searchURI = buildSearchRequest(startTime, endTime, 1);
+
+            String response = getZendeskRestClient().doGet(searchURI, task, Exec.isPreview());
+
+            JsonNode json = ZendeskUtils.parseJsonObject(response);
+
+            if (!ZendeskUtils.isNull(json)) {
+                int totalRecords = json.get("count").asInt();
+                int totalPages = totalRecords > MAXIMUM_TOTAL_RECORDS
+                    ? MAXIMUM_TOTAL_PAGES
+                    : slicePages(totalRecords);
+
+                logger.info(String.format("Fetching from '%s' to '%s' with '%d' pages and '%d' records",
+                    startTime, endTime, totalPages,  totalRecords));
+
+                final String lastEndTime = endTime;
```

Comment:
Because of the streaming use here: https://github.com/treasure-data/embulk-input-zendesk/pull/65/files/908f8535d3c09b06f31c7fdeddbb5702d4c0d0fa#diff-6bf3ea9861bae84404dacb2bcbb3e569R86. need final variable but we also update the value of `end_time` here: 
https://github.com/treasure-data/embulk-input-zendesk/pull/65/files/908f8535d3c09b06f31c7fdeddbb5702d4c0d0fa#diff-6bf3ea9861bae84404dacb2bcbb3e569R97

Reply:
got it, if so how about the change:
```
logger.info(String.format("Fetching from '%s' to '%s' with '%d' pages and '%d' records",
                    startTime, endTime, totalPages,  totalRecords));

                if (totalPages > 1) {
                    // endTime local variable here is required only because of Lambda requires final variable
                    final String tempEndTime = endTime;

                    // Can't start from 0 because page 0 and 1 return the same data, and according to document, page start from 1
                    IntStream.range(1, totalPages + 1).parallel().forEach(page -> fetchData(startTime, tempEndTime, page, recordImporter));
                }
                else {
                    // in case totalPages 1, IntStream not work
                    fetchData(startTime, endTime, 1, recordImporter);
                }
```

It minimizes the local variable scope to the place where it is needed to get rid of the confusion.